### ISSUE TITLE
Vue 3 timing issue with select control example for components (in this case @carbon/icons-vue icon)

### ISSUE THREAD

**Issue by lee-chase** [storybookjs/storybook#13861]
**Describe the bug**

The component is displayed if I refresh the browser, but not when selecting a story for a second time.

**To Reproduce**
Steps to reproduce the behaviour:

1. Create a simple component expecting an Icon component (using @carbon/vue-icons) e.g. a button
2. Write your story along the lines of the icon example here https://storybook.js.org/docs/vue/essentials/controls
3. View in the storybook.
4. Change the story a few times and note the icon stops being displayed

**Expected behavior**
Icon/Component should be displayed as expected.

**Screenshots**
Icon no longer displayed after changing story from Primary to Secondary and back.
![image](https://user-images.githubusercontent.com/15086604/107499217-9687ae80-6b8c-11eb-9ba4-589a11b0ae00.png)
![image](https://user-images.githubusercontent.com/15086604/107499245-a0111680-6b8c-11eb-8f9a-6e37b92c574d.png)
![image](https://user-images.githubusercontent.com/15086604/107499283-ab644200-6b8c-11eb-9373-a95704fd48d3.png)

**Code snippets**
FAILING VERSION
```js
import {
  Bee20,
  Carbon20,
  Watson20,
  IbmCloud20,
  EdtLoop20,
  IbmSecurity20,
} from '@carbon/icons-vue';

const icons = {
  Bee20,
  Carbon20,
  Watson20,
  IbmCloud20,
  EdtLoop20,
  IbmSecurity20,
};

export default {
  title: 'Components/CvIconButton',
  component: CvIconButton,
  argTypes: {
    icon: {
      control: {
        type: 'select',
        options: Object.keys(icons),
        value: 'Bee20',
      },
    },
  },
};

const Template = (args, { argTypes }) => {
  args.icon = icons[args.icon];
  return {
    props: Object.keys(argTypes),
    components: { CvIconButton },
    template: `<cv-icon-button v-bind="$props"/>`,
  };
};

export const Primary = Template.bind({});
export const Secondary = Template.bind({});
export const Tertiary = Template.bind({});
```

WORK AROUND - changed template to
```js
const Template = (args, { argTypes }) => {
  return {
    props: Object.keys(argTypes),
    components: { CvIconButton },
    template: `<cv-icon-button :data-test="label" v-bind="$props" :icon="icon"/>`,
    setup(props) {
      const icon = shallowRef(icons[props.icon]);

      if (icon.value === undefined) {
        // assigning an icon seem sto need a kick
        setTimeout(() => {
          icon.value === icons[props.icon];
        }, 1);
      }

      return { icon };
    },
  };
};
```

**System**
Please paste the results of `npx sb@next info` here.

$npx sb@next info

Environment Info:

  System:
    OS: macOS 11.1
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  Binaries:
    Node: 14.15.3 - ~/.nvm/versions/node/v14.15.3/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.14.9 - ~/.nvm/versions/node/v14.15.3/bin/npm
  Browsers:
    Chrome: 88.0.4324.150
    Edge: 88.0.705.63
    Safari: 14.0.2
  npmPackages:
    @storybook/addon-actions: ^6.2.0-alpha.24 => 6.2.0-alpha.24 
    @storybook/addon-essentials: ^6.2.0-alpha.24 => 6.2.0-alpha.24 
    @storybook/addon-links: ^6.2.0-alpha.24 => 6.2.0-alpha.24 
    @storybook/vue3: ^6.2.0-alpha.24 => 6.2.0-alpha.24 
    
**Additional context**

**Comment by phated** [comment]
@lee-chase I think this might be because I used a `shallowRef` to track the args. I didn't know it would be a common case to mutate the props inside a story.

I think the solution would be to make the args `reactive`, as was the case in the original Vue 3 PR. WDYT?

**Comment by phated** [comment]
Hmmmm, I just wrote a more simplified version of this in the vue3 example and it seems to be working without any changes to the vue3 implementation. I'm trying to figure out the differences between my simplified version and your example above (I might create a branch for you to check out).

**Comment by phated** [comment]
@lee-chase do you have a branch or something I can try out? I tried searching https://github.com/carbon-design-system/carbon-components-vue but I can't seem to find something like your reference above.

**Comment by phated** [comment]
Ah, I found it at https://github.com/carbon-design-system/carbon-components-vue/blob/master/packages/v3/src/components/CvButton/CvButton.stories.js

**Comment by phated** [comment]
Ah, @lee-chase I think I figured out what the problem is, if I may:
1. The first time through the render, your arg is passed to the story as a string, which you then convert to an object and re-assign to your args.
2. Args are the same reference between page refreshes, so the next time it tries to render the story, your arg is already an object.
3. Trying to lookup an object as the key into your icons object doesn't work and just returns `undefined`, so the component doesn't have an icon to render.

**Comment by phated** [comment]
I changed your story to this, and it works:

```js
const Template = (args, { argTypes }) => {
  args.icon = typeof args.icon === 'string' ? icons[args.icon] : args.icon;
  return {
    props: Object.keys(argTypes),
    components: { CvIconButton },
    template: `<cv-icon-button v-bind="$props"/>`,
  };
};
```

**Comment by phated** [comment]
I _believe_ it is by design that inside the preview an `args` instance is always the same object, so values adjusted by controls persist between navigation. Maybe @shilman can correct me if I'm wrong about that.

If my understanding is correct, I don't think there's anything for us to change here ❓

**Comment by phated** [comment]
I chatted with @shilman and he described the use case that you are trying to support here. I'm going to create an example of how this override can be done with Vue 3.

**Comment by phated** [comment]
I just put up a PR (linked above) that shows how to override props in a story without mutating the props.

**Comment by lee-chase** [comment]
Does this example also need updating @phated? I believe it follows the same pattern I used above.

https://storybook.js.org/docs/vue/essentials/controls#fully-custom-args

**Comment by lee-chase** [comment]
P.S. Thanks for the nudge on the template. I think I prefer the following as a Vue pattern.

```
const Template = (args, { argTypes }) => {
  return {
    props: Object.keys(argTypes),
    components: { CvIconButton },
    template: `<cv-icon-button v-bind="$props" :icon="icons[icon]"/>`,
    setup() {
      return { icons };
    },
  };
};
```

**Comment by shilman** [comment]
> Does this example also need updating @phated? I believe it follows the same pattern I used above.
> storybook.js.org/docs/vue/essentials/controls#fully-custom-args

@phated @jonniebigodes we probably need a `storybook.js.org/docs/vue3/essentials/controls#fully-custom-args` OR we need some other way to update the snippets

**Comment by phated** [comment]
I think we need separate docs, but also need to update the vue docs to remove that mutation, since that causes unintended side-effects. I think vue 2 might support the same use-case with `data` and I tested the `setup` method some more in @lee-chase's example and it seems to work when mapping (no timeout needed):
```
setup(props) {
  return {
    icon: icons[props.icon]
  }
}
```

**Comment by jonniebigodes** [comment]
@phated and @shilman i've been doing some testing and we could take the same approach as react js and ts

we could call the snippet `your-component.vue3.js.mdx` for vue 3 and `your-component.vue2.js.mdx` and you get the following in the docs:

![example-vue2-vue3](https://user-images.githubusercontent.com/22988955/108134423-1d1d1e00-70ae-11eb-9a60-ddb8559d947d.png)

**Comment by shilman** [comment]
Great idea @jonniebigodes ❤️

**Comment by shilman** [comment]
closed by #13981 i believe. @phated ?

**Comment by phated** [comment]
Yep!
